tracer/src/Datadog.Trace/IAST/Propagation/StringModuleImpl.cs

```diff
@@ -0,0 +1,303 @@
+
+            TaintedObjects taintedObjects = iastContext.GetTaintedObjects();
+
+            Range[]? ranges = null;
+            int length = 0;
+            foreach (var parameterAsObject in parameters)
+            {
+                var currentParameter = parameterAsObject?.ToString();
+                if (!CanBeTainted(currentParameter))
+                {
+                    continue;
+                }
+
+                var taintedParameter = GetTainted(taintedObjects, currentParameter);
+                if (taintedParameter != null)
+                {
+                    if (ranges == null)
+                    {
+                        if (taintedParameter != null)
+                        {
+                            if (length == 0)
+                            {
+                                ranges = taintedParameter.Ranges;
+                            }
+                            else
+                            {
+                                ranges = new Range[taintedParameter!.Ranges!.Length];
+                                Ranges.CopyShift(taintedParameter!.Ranges, ranges, 0, length);
+                            }
+                        }
+
+                        length += currentParameter!.Length;
+                        continue;
+                    }
+
+                    ranges = Ranges.MergeRanges(length, ranges, taintedParameter.Ranges);
+                }
+
+                length += currentParameter!.Length;
+            }
+
+            if (ranges != null)
+            {
+                taintedObjects.Taint(result, ranges);
+            }
+        }
+        catch (Exception err)
+        {
+            Log.Error(err, "StringModuleImpl.OnstringConcat(IEnumerable) exception {Exception}", err.Message);
+        }
+
+        return result;
+    }
+
+    public readonly struct StringConcatParams
+    {
+        public readonly string? P0;
+        public readonly string? P1;
+        public readonly string? P2;
+        public readonly string? P3;
+        public readonly string? P4;
+        public readonly int ParamCount;
+
+        public StringConcatParams(string? p0, string? p1, string? p2)
+        {
+            P0 = p0;
+            P1 = p1;
+            P2 = p2;
+            ParamCount = 3;
+        }
+
+        public StringConcatParams(string? p0, string? p1, string? p2, string? p3)
+        {
+            P0 = p0;
+            P1 = p1;
+            P2 = p2;
+            P3 = p3;
+            ParamCount = 4;
+        }
+
+        public StringConcatParams(string? p0, string? p1, string? p2, string? p3, string? p4)
+        {
+            P0 = p0;
+            P1 = p1;
+            P2 = p2;
+            P3 = p3;
+            P4 = p4;
+            ParamCount = 5;
+        }
+
+        public string? this[int index]
+        {
+            get
+            {
+                if (index < 0 || index >= ParamCount)
+                {
+                    throw new IndexOutOfRangeException();
```

Comment:
Throwing avoids inlining of this indexer, if you want an inlining opportunity
use https://github.com/DataDog/dd-trace-dotnet/blob/eb4515a4dc6bd3d04f5211eef87a10318013fc4a/tracer/src/Datadog.Trace/Util/ThrowHelper.cs class to throw in performance scenarios.

Reply:
I have updated the code. Thanks for the tip!!!